solr/core/src/java/org/apache/solr/security/AllowListUrlChecker.java

```diff
@@ -0,0 +1,215 @@
+     *                    An empty list means there is no explicit allow-list of URLs, in this case no
+     *                    URL is allowed unless a {@link ClusterState} is provided in
+     *                    {@link #checkAllowList(List, ClusterState)}.
+     * @throws MalformedURLException If an URL is invalid.
+     */
+    public AllowListUrlChecker(List<String> urlAllowList) throws MalformedURLException {
+        hostAllowList = parseHostPorts(urlAllowList);
+    }
+
+    /**
+     * Creates a URL checker based on the {@link NodeConfig} property to configure the allowed URLs.
+     */
+    public static AllowListUrlChecker create(NodeConfig config) {
+        if (Boolean.getBoolean(DISABLE_URL_ALLOW_LIST)) {
+            return AllowListUrlChecker.ALLOW_ALL;
+        } else if (System.getProperty("solr.disable.shardsWhitelist") != null) {
+            log.warn("Property 'solr.disable.shardsWhitelist' is deprecated, please use '{}' instead.", DISABLE_URL_ALLOW_LIST);
+        }
+        try {
+            return new AllowListUrlChecker(config.getAllowUrls());
+        } catch (MalformedURLException e) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                    "Invalid URL syntax in '" + URL_ALLOW_LIST + "' configuration: " + config.getAllowUrls(), e);
+        }
+    }
+
+    /**
+     * @see #checkAllowList(List, ClusterState)
+     */
+    public void checkAllowList(List<String> urls) throws MalformedURLException {
+        checkAllowList(urls, null);
+    }
+
+    /**
+     * Checks that the given URLs are present in the configured allow-list or in the provided {@link ClusterState}
+     * (in case of cloud mode).
+     *
+     * @param urls         The list of urls to check.
+     * @param clusterState The up to date {@link ClusterState}, can be null in case of non-cloud mode.
+     * @throws MalformedURLException If an URL is invalid.
+     * @throws SolrException         If an URL is not present in the allow-list or in the provided {@link ClusterState}.
+     */
+    public void checkAllowList(List<String> urls, @Nullable ClusterState clusterState) throws MalformedURLException {
+        Set<String> clusterHostAllowList = clusterState == null ? Collections.emptySet() : clusterState.getHostAllowList();
+        for (String url : urls) {
+            String hostPort = parseHostPort(url);
+            if (!clusterHostAllowList.contains(hostPort) && !hostAllowList.contains(hostPort)) {
+                throw new SolrException(SolrException.ErrorCode.FORBIDDEN, "URL " + url +
+                        " is neither a live node of the cluster nor in the configured '" +
+                        URL_ALLOW_LIST + "' " + hostAllowList);
+            }
+        }
+    }
+
+    /**
+     * Whether this checker has been created with a non-empty allow-list of URLs.
+     */
+    public boolean hasExplicitAllowList() {
+        return !hostAllowList.isEmpty();
```

Comment:
Woah; 4-space indent.  Please fix this and change your IDE settings.  Use "Google Java Format" IntelliJ plugin, which may help and is practically required on the Lucene side any way.

Reply:
Strange my IDE lost some settings. I'll install this plugin, thanks!